README
napple1 is an Apple 1 emulator using ncurses, ported from the SDL version 
Pom1 emulafor. If you have some questions, please report an issue on github 
or mail to mail@nobuh.live-on.net

How to install on cygwin
=========================
0. Setup cygwin (base) and make, gcc4, libncurses-dev
   and follow Linux procedure. 

How to install on Linux
========================
1. Download latest source package nappple1-<n>.<n>.tar.gz 
   and extract it.

2. Build
	cd napple1/src
	make
	cd ..
3. Run 
	./napple1

ROM Directory
======================
napple1 contain basic.rom and monitor.rom files on
the ./rom directory. However, If you want to put 
basic.rom and monitor.rom on another directory. 
Please use ROMDIR environmental variable.

For example, if you want to put basic.rom and monitor.rom 
on current directory, the command is:

    ROMDIR=. ./napple1

Default value of the ROMDIR is ./rom 
You can use an absolute path for the ROM Directory.

    ROMDIR=/home/username/my_rom_files ./napple1

How to uninstall
================
	cd napple1/src
	make clean

How to use programs on napple1
==============================

- Key map

In normal terminals, if you press 'D' key, it will be echoed as 'd' on screen.
And press 'Shift + D' prints 'D' on screen. Shift key works to flip caps. 

However, napple1, if you press 'D' key, it will be echoed 'D'. All letters 
you can type is upper case only. Because original apple1 has a limit to do so.
And shift key works like a control key in napple1.

- Cold start Apple 1 Basic

type E000R<enter>

- Warm start Apple 1 Basic

type E2B3R<enter>

- Emulator commands

Command      Key        Description
--------------------------------------------------------------------
Basic load   Shidt + B  Load rom/basic.rom to ram
Dump core    Shift + D  Dump memory to core/_file name_ 
Load core    Shift + L  Load memory from core/_file name_ 
Reset        Shift + R  Reset the emulator
Hard reset   Shidt + H  Reset & Clear memory
Quit         Shidt + Q  Quit the emulator
Mode         Shift + M  Mode change 8KB & 32KB ram mode

- Load / Save programs

To load or save program, please load or dump entire 64kb memory.

- napple1 memory map in memory.c

/*
 * Memory maap of napple1
 * 
 * Apple I 8K mode & napple1 64K mode, common usage
 * -----------------------------------------------------------------------
 * $0000           Ststem & User Space 
 * $0200 - $027F   Input Buffer used by monitor
 * $0FFF
 * -----------------------------------------------------------------------
 * $D010	KBD	Keyboard input register. b7 is always 1 by hardware.
 * 		Read KBD will automatcically clear KBDCR's b7.
 * $D011	KBDCR	When key is pressed, b7 is set by hardware.
 * $D012	DSP	Bit b6..b0 is output character for the terminal.
 *	        Writing to DSP will set b7 by hardware.
 *              The termianl clear b7 after the character is accepted.
 * $D013	DSPCR	Unused.
 * -----------------------------------------------------------------------
 * $E000        Apple I Integer BASIC
 * $E2B3        Re-entry address
 * $EFFF
 * -----------------------------------------------------------------------
 * $FF00        Monitor
 * $FFEF	Echo
 * $FFFF	
 * ----------------------------------------------------------------------- 
 */
/* Apple I 8K mode memory map
 * --------------------------------- 
 * Start Type
 * addr
 * --------------------------------- 
 * $0000 4KB RAM
 * $1000 unused
 * $D010 Display and Keyboard I/O
 * $D014 unused
 * $E000 4KB RAM
 * $F000 unused
 * $FF00 256B ROM^ (Woz Monitor)
 * ---------------------------------
 * ^ ROM can be written by Load core  
 */
/* napple I 32K mode memory map
 * --------------------------------- 
 * Start Type
 * addr
 * --------------------------------- 
 * $0000 32K RAM
 * $8000 unused
 * $D010 Display and Keyboard I/O
 * $D014 unused
 * $E000 8K ROM^ 
 * ---------------------------------
 * ^ ROM can be written by Load core  
 */




CHANGELOG
CHANGELOG
=========

2.4.0 mtleblanc's contribution:
- Indirect X zero page addressing should always wrap and stay on zero page
- BRK should disable interrupts
- ADC in decimal mode should carry between nibbles based on adjusted,
  not original values.
- Writes of status register to the stack should always set M and should
  set B on BRK/PHP and ~B on IRQ/NMI
- Given the above, status register does not need to track M or B bits,
  removed from initialization.
- Found and tested against
  https://github.com/Klaus2m5/6502_65C02_functional_tests

2.3.0
- ROM can be loaded from the directory of ROMDIR environment variable.

2.2.1
- Fix leaking memory on fd of loadBasic() function
- Fix some coding style issue in memory.c and msgbuf.c 

2.2
- Add Cygwin (with cygwin base, make, gcc4 and libncurses-dev) support
- Fix Hung after Hardreset. 

2.1
- If a loading core has a value on the 0xF000 area, the mode
  should be better change to 32K mode. 
- Update SAMPLE_PROGRAMS file. 

2.0
- Remove TODO list. 
- Synchronize video output for ideal 60 Hz screen refresh.
- Remove terminalSpeed configuration. Now it is 60Hz only.
- Remove replica1 config in napple1.cfg and No more napple1.cfg.
- Mode change command filp 8K memory mode & 32K memory mode
- Delete unused charmap.rom
- Rename core/STARTREK to core/startrek

1.0
- Filename input for Load core.
- Confirmation to load basic to ram.
- Command list on the message buffer.
- Remove default core and basic.rom loading function.
- TODO file added.

0.6 
- March 29, 2010 
- Modify makefile as using macros for common options.
- Add #define _XOPEN_SOURCE 500 to prevent compile warning of usleep. 
- Add filename input feature for 'D'ump core.  

0.5 
- March 25, 2010
- Fix time tic in m6502. To get milli sec current time, use gettimeofday.
- Wall option at compile. So, several fix in m6502.c to get rid of warnings.  

0.4
- March 25, 2010
- Fix some descriptions in READE and SAMPLE_PROGRAMS
- Use independant ncurses windows for main screen and message buffer, 
  instead of a single stdscr. The new screen layoht is following.
  During loading rom and core, printing some messages to the buffer.

	+-----------------------------+
	|(0,0)                 (0, 40)|
	|          screen             |
	|                             |
	|(23,0)                (23,40)|
	+-----------------------------+
	|(24,0) message buffer (24,40)|
	+-----------------------------+

0.3
- March 21, 2010
- Update README 
- Allow smaller lines/colums terminal

0.2
- March 18, 2010
- Delete automatic save memory function during quit.
- Shift + D "Dump core"
- Dump & Load core on file core/dumped
- Add sample core file in core directory
	
0.1
- March 16, 2010
- Fixed old program name nPom1 to napple1.
- napple1 now includes Pom1 rom.
- if napple1 started with normal mode, and if core file 
  isn't existing, load rom/basic.rom to ram 0xE000. 

0.0 
- March 15, 2010
- Derived from Pom1. First alpha release


src/memory.c
/* napple1 ncurses Apple 1 emulator
 * Copyright (C) 2008 Nobu Hatano
 *
 * Pom1 Apple 1 Emulator
 * Copyright (C) 2000 Verhille Arnaud
 * Copyright (C) 2006 John D. Corrado
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/*
 * Memory maap of napple1
 * 
 * Apple I 8K mode & napple1 64K mode, common usage
 * -----------------------------------------------------------------------
 * $0000           Ststem & User Space 
 * $0200 - $027F   Input Buffer used by monitor
 * $0FFF
 * -----------------------------------------------------------------------
 * $D010	KBD	Keyboard input register. b7 is always 1 by hardware.
 * 		Read KBD will automatcically clear KBDCR's b7.
 * $D011	KBDCR	When key is pressed, b7 is set by hardware.
 * $D012	DSP	Bit b6..b0 is output character for the terminal.
 *	        Writing to DSP will set b7 by hardware.
 *              The termianl clear b7 after the character is accepted.
 * $D013	DSPCR	Unused.
 * -----------------------------------------------------------------------
 * $E000        Apple I Integer BASIC
 * $E2B3        Re-entry address
 * $EFFF
 * -----------------------------------------------------------------------
 * $FF00        Monitor
 * $FFEF	Echo
 * $FFFF	
 * ----------------------------------------------------------------------- 
 */
/* Apple I 8K mode memory map
 * --------------------------------- 
 * Start Type
 * addr
 * --------------------------------- 
 * $0000 4KB RAM
 * $1000 unused
 * $D010 Display and Keyboard I/O
 * $D014 unused
 * $E000 4KB RAM
 * $F000 unused
 * $FF00 256B ROM^ (Woz Monitor)
 * ---------------------------------
 * ^ ROM can be written by Load core  
 */
/* napple I 32K mode memory map
 * --------------------------------- 
 * Start Type
 * addr
 * --------------------------------- 
 * $0000 32K RAM
 * $8000 unused
 * $D010 Display and Keyboard I/O
 * $D014 unused
 * $E000 8K ROM^ 
 * ---------------------------------
 * ^ ROM can be written by Load core  
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "pia6820.h"
#include "memory.h"
#include "screen.h"
#include "msgbuf.h"

#define MEMMAX 0xFFFF
#define FNAME_LEN_MAX 1024

static unsigned char mem[65536];
static int mode = 8; /* 8 = Apple I 8K mode, 32 = napple1 32K mode */

char rombasic[FNAME_LEN_MAX];
char rommonitor[FNAME_LEN_MAX];

void flipMode(void)
{
	if (mode == 8)
		mode = 32;
	else 
		mode = 8;

	/* update message buffer */
	print_msgbuf("");
}

int memMode(void) 
{
	return mode;
}

void loadBasic(void)
{
	FILE *fd = fopen(rombasic, "rb");
	char input[MSG_LEN_MAX +1];
	
	if (!fd) {
		gets_msgbuf("Failed to open basic.rom", input);
		return;
	}

	gets_msgbuf("Load basic.rom to ram? y/n: ", input);
	if (input[0] == 'y') {
		size_t s = fread(&mem[0xE000], 1, 4096, fd);
		if (s) {
			gets_msgbuf("Load completed: ", input);
		} else {
			gets_msgbuf("Load failed: ", input);
		}
	}

	fclose(fd);
	return;
}

int loadMonitor(void)
{
	FILE *fd = fopen(rommonitor, "rb");

	if (fd) {
		fread(&mem[0xFF00], 1, 256, fd);
		fclose(fd);
	}
	else{
		return 0;
	}

	return 1;
}

void resetMemory(void)
{
	if (memMode() > 8)
		memset(mem, 0, 0xE000); /* rom is starting from 0xE000 */
	else
		memset(mem, 0, 0x10000 - 256); /* rom is within tail 256b */
}

unsigned char memRead(unsigned short address)
{
	if (address == 0xD013)
		return readDspCr();
	if (address == 0xD012)
		return readDsp();
	if (address == 0xD011)
		return readKbdCr();
	if (address == 0xD010)
		return readKbd();

	return mem[address];
}

void memWrite(unsigned short address, unsigned char value)
{
	if (address < 0x1000)	
		mem[address] = value;
	else if (address < 0x8000 && (memMode() > 8) )
		mem[address] = value;
	else if (address == 0xD013)
		writeDspCr(value);
	else if (address == 0xD012)
		writeDsp(value);
	else if (address == 0xD011)
		writeKbdCr(value);
	else if (address == 0xD010)
		writeKbd(value);
	else if (address >= 0xE000 && address < 0xF000 && memMode() == 8)
		mem[address] = value;
	else
		;

	return;
}

void dumpCore(void)
{
	int i;
	FILE *fd;
	char input[MSG_LEN_MAX +1];
	char corename[5 + MSG_LEN_MAX +1]; /* 'core/' + input string */

	gets_msgbuf("Dump core. Filename: ", input);
	sprintf(corename, "core/%s", input);

	fd = fopen(corename, "w");
	for (i = 0; i <= MEMMAX; i++)
		fputc(mem[i], fd);
	fclose(fd);
	gets_msgbuf("Dump core completed: ", input);
}

int loadCore(void)
{
	FILE *fd;
	char input[MSG_LEN_MAX +1];
	char corename[5 + MSG_LEN_MAX +1]; /* 'core/' + input string */
	size_t s = 0;
	unsigned char buf[65536];
	int i;

	gets_msgbuf("Load core. Filename: ", input);
	sprintf(corename, "core/%s", input);

	fd = fopen(corename, "r");
	if (fd) {
		s = fread(&buf[0], 1, MEMMAX+1, fd);
		fclose(fd);
	}
	if (s) { 
		gets_msgbuf("Load core completed: ", input);
	} else {
		gets_msgbuf("Failed to open core file: ", input);
		return 0;
	}

	/* 0xF000 is unused area of 8K mode or
	 * ROM area of 32K mode. So,  if 0xF000 has a value,
	 * The mode should better change to 32K mode.
	 */
	if ((buf[0xF000] != 0) && (memMode() == 8)) {
		flipMode();
	}

	if (memMode() == 8) {
		for (i = 0;      i <= 0x0FFF; i++) mem[i] = buf[i];
		for (i = 0xE000; i <= 0xEFFF; i++) mem[i] = buf[i];
		for (i = 0xFF00; i <= 0xFFFF; i++) mem[i] = buf[i];
	} else {
		for (i = 0;      i <= 0x7FFF; i++) mem[i] = buf[i];
		for (i = 0xE000; i <= 0xFFFF; i++) mem[i] = buf[i];
	}
	return 1;
}

/* set ROM file name using ROMDIR env variable
 * default path is ./rom 
 * need to be called before loadBasic() and loadMonitor()
 */
void setRomFiles(void)
{
    char env[FNAME_LEN_MAX];
    char *p;

    strcpy(rombasic, "rom/basic.rom");
    strcpy(rommonitor, "rom/monitor.rom");

    p = env;
    if (getenv("ROMDIR")) {
        p = getenv("ROMDIR");
        sprintf(rombasic, "%s/basic.rom", p);
        sprintf(rommonitor, "%s/monitor.rom", p);
    } 
}


